compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -431,10 +431,14 @@ object Erasure {
       val implParamTypes = implType.paramInfos
       val implResultType = implType.resultType
       val implReturnsUnit = implResultType.classSymbol eq defn.UnitClass
-      // The SAM that this closure should implement
-      val SAMType(sam) = lambdaType: @unchecked
-      val samParamTypes = sam.paramInfos
-      val samResultType = sam.resultType
+      // The SAM that this closure should implement.
+      // At this point it should be already guaranteed that there's only one method to implement
+      val Seq(unerasedSam) = lambdaType.possibleSamMethods
+      // We're now in erasure so the alternatives will have erased types
+      val Seq(erasedSam: MethodType) = unerasedSam.symbol.alternatives.map(_.info)
```

Comment:
To avoid this complication I suggest changing `possibleSamMethods` to do `.map(_.current)` at the end so the denotations always correspond to the current phase.